[Bill Xiao] iP

xiaobill8 opened the pull request:
# DukePro
> “Your mind is for having ideas, not holding them.” – David Allen [source](https://dansilvestre.com/productivity-quotes/)

DukePro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* FAST SUPER FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is FREE!

Features:

- [x] Managing tasks

- [ ] Managing deadlines (coming soon)

- [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the main method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(Duke.class, args);
    }
}
```

Commit:
Add Gradle support

xiaobill8 committed:
Completed Level 1. Greet, Echo, Exit

xiaobill8 committed:
Completed Level 2. Add, List

xiaobill8 committed:
Completed Level. 3 Mark as Done, Extension: A-Classes

xiaobill8 committed:
Completed Level 4. ToDos, Events, Deadlines. Added Inheritance for multiple task types.

xiaobill8 committed:
Completed Automated Text UI Testing

xiaobill8 committed:
Completed Level 5. Handle Errors. Implemented A-Exceptions.

xiaobill8 committed:
Completed Level 6. Delete

xiaobill8 committed:
Completed Level 7. Save

xiaobill8 committed:
Completed Level 8. Date and Times

xiaobill8 committed:
Merge branch 'branch-level-7'

xiaobill8 committed:
Aligned folder structure.

xiaobill8 committed:
Removed unnecessary imports

xiaobill8 committed:
Merge branch branch-level-8

xiaobill8 committed:
Fixed issues from merging branch-Level-8

xiaobill8 committed:
Refactored code to use more OOP

xiaobill8 committed:
Refactored code with packages

xiaobill8 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

xiaobill8 committed:
Completed A-JUnit: Add JUnit Tests

xiaobill8 committed:
Completed A-Jar: Create a JAR File

amresh2000 commented on src/main/java/duke/models/TaskList.java:
```diff
@@ -0,0 +1,7 @@
+import java.util.ArrayList;
+
+public class TaskList extends ArrayList<Task> {
+    private final ArrayList<Task> list = new ArrayList<>();
```
A more meaningful name could have been used

amresh2000 commented on src/test/java/duke/TaskListTest/TaskListTest.java:
```diff
@@ -0,0 +1,20 @@
+    TaskList list = new TaskList();
+
+    @Test
+    public void addTaskTest() {
```
Good naming style for methods used throughout the code

amresh2000 reviewed (commented):
Overall I found your code easy to read for the most part except for some minor suggestions. Consider it a job well done 😃

xiaobill8 committed:
Completed A-JavaDoc: JavaDoc

xiaobill8 committed:
Completed A-CodingStandard: Follow the Coding Standard

xiaobill8 committed:
Completed Level-9.Find

xiaobill8 committed:
Merge branch 'branch-A-JavaDoc'

xiaobill8 committed:
Merge branch 'branch-A-CodingStandard'

xiaobill8 committed:
Merged branch branch-Level-9

leirdas commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,78 @@
+
+    private final LocalStorage storage;
+    private final Parser parser;
+    private TaskList list;
```
Perhaps the TaskList attribute could have a better name such as tasks. In the coding standards, plural form should be used on names representing a collection of objects, maybe you could look at renaming it to a plural form.

leirdas commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,78 @@
+            String[] fullCommand = parser.parseFullCommand(input);
+
+            try {
+                switch (command) {
+                    case BYE:
+                        ByeHandler.handle();
+                        break;
+                    case LIST:
+                        ListHandler.handle(list);
+                        break;
+                    case TODO:
+                        TodoHandler.handle(list, fullCommand[1]);
+                        break;
+                    case DEADLINE:
+                        DeadlineHandler.handle(list, fullCommand[1]);
+                        break;
+                    case EVENT:
+                        EventHandler.handle(list, fullCommand[1]);
+                        break;
+                    case MARK:
+                        MarkHandler.handle(list, fullCommand[1]);
+                        break;
+                    case UNMARK:
+                        UnmarkHandler.handle(list, fullCommand[1]);
+                        break;
+                    case DELETE:
+                        DeleteHandler.handle(list, fullCommand[1]);
+                        break;
+                    default:
```
The coding standards state that there is no indentation for case clauses.

leirdas commented:
Overall, I found the card easy to read and I like how you split the different classes that clearly handle specific functions. Other than that, there are a few coding violations to take note of.

leirdas reviewed (commented):
Overall, I found the card easy to read and I like how you split the different classes that clearly handle specific functions. Other than that, there are a few coding violations to take note of.

xiaobill8 committed:
Completed 10.GUI

xiaobill8 committed:
Complete A-Assertions

Java's assert feature documents important assumptions that should
hold at various points in the code.

Assertions are added to ensure that lists contain
the correct number of elements when they are indexed.

xiaobill8 committed:
Complete A-CodeQuality

Critically examine the code and refactor to improve code quality
and readability where necessary.

Supporting DateTime in Event's At timing and Deadline's By timing
allows for additional granularity of timestamp.

The supported DateTime format for creating Event/Deadline is
"yyyy-MM-dd HH:mm"

xiaobill8 committed:
Complete B-RecurringTasks

Support Recurring Events to avoid creating repeated Events.

Recurring events support 3 different recurring intervals, Day, Week and Month.
When a recurring event is marked as done, a new Event at Date + Interval
is created automatically.
The marked event remains in the TaskList and can be deleted with no side effects.
When the subsequently created event is marked done, the process is repeated
until the unmarked recurring task is deleted.

The new format for creating recurring Event is
"event <description> /at <Date> [/r <Interval>]"
where [] denotes optional for recurring events only

xiaobill8 committed:
Merge pull request #2 from xiaobill8/branch-A-Assertions

Merge branch-A-Assertions

xiaobill8 committed:
Merge pull request #3 from xiaobill8/branch-A-CodeQuality

Complete A-CodeQuality

xiaobill8 committed:
Merge branch-B-RecurringTasks

xiaobill8 committed:
Merge pull request #4 from xiaobill8/branch-B-RecurringTasks

Complete B-RecurringTasks

xiaobill8 committed:
Improve Recurring Task Logic

xiaobill8 committed:
Complete A-UserGuide

xiaobill8 committed:
Update Recurring Task logic

xiaobill8 committed:
Completed A-BetterGui

Fixed support for text wrapping when displayed response
is longer than the default height.

xiaobill8 committed:
Update Testing

Updated JUnit tests with recurring tasks.

xiaobill8 committed:
Complete A-CheckStyle

xiaobill8 committed:
Complete A-CheckStyle

xiaobill8 committed:
Complete A-MoreErrorHandling

xiaobill8 committed:
Minor wording changes

xiaobill8 committed:
Error Handling Fixes

xiaobill8 committed:
Minor NullPointerException Bug Fix

xiaobill8 committed:
Fixed import *
